### PR TITLE
[TRUS-383] - Agent logged in front-end microservice has HMRC-AS-AGENT enrolment

### PR DESCRIPTION
**Pull request by jassalrichy**
[TRUS-383] - Agent logged in front-end microservice has HMRC-AS-AGENT enrolment

### CODE REVIEW

File: app/controllers/actions/IdentifierAction.scala [hunk 1]
```diff
@@ -35,21 +37,26 @@ class AuthenticatedIdentifierAction @Inject()(
...
     } recover {
-      case _: NoActiveSession =>
-        Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl)))
-      case _ =>
-        Redirect(routes.UnauthorisedController.onPageLoad())
-
+      case ex: NoActiveSession => Redirect(config.loginUrl, Map("continue" -> Seq(config.loginContinueUrl)))
+      case ex: InsufficientEnrolments => Redirect(routes.UnauthorisedController.onPageLoad)
+      case ex: InsufficientConfidenceLevel => Redirect(routes.UnauthorisedController.onPageLoad)
+      case ex: UnsupportedAuthProvider => Redirect(routes.UnauthorisedController.onPageLoad)
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/controllers/actions/IdentifierAction.scala]: @manishsw Yes

File: app/controllers/actions/IdentifierAction.scala [hunk 2]
```diff
@@ -35,21 +37,26 @@ class AuthenticatedIdentifierAction @Inject()(
...
-
-    authorised().retrieve(Retrievals.internalId) {
-      _.map {
-        internalId => block(IdentifierRequest(request, internalId))
-      }.getOrElse(throw new UnauthorizedException("Unable to retrieve internal Id"))
+    val hmrcAgentEnrolmentKey = "HMRC-AS-AGENT"
+    authorised().retrieve(Retrievals.internalId and Retrievals.affinityGroup and Retrievals.allEnrolments) {
+      case Some(internalId) ~ Some(Agent) ~ enrolments => {
+        if (enrolments.getEnrolment(hmrcAgentEnrolmentKey).nonEmpty)
+          block(IdentifierRequest(request, internalId))
```
[the remark this answers is not in the document]
Comment (jassalrichy) [reply app/controllers/actions/IdentifierAction.scala]: @manishsw @adamconder Not part of this ticket. Will need to be done in another ticket.

File: test/controllers/actions/AuthActionSpec.scala [hunk 3]
```diff
@@ -35,8 +35,76 @@ class AuthActionSpec extends SpecBase {
...
+        val authAction = new AuthenticatedIdentifierAction(new FakeAuthConnector(authRetrievals(agentAffinityGroup, noEnrollment)), frontendAppConfig, bodyParsers)
+        val controller = new Harness(authAction)
+        val result = controller.onPageLoad()(fakeRequest)
+        status(result) mustBe SEE_OTHER
+        redirectLocation(result) mustBe Some(controllers.routes.CreateAgentServicesAccountController.onPageLoad().url)
+      }
+
+    }
+
+    "Agent user has correct enrolled in Trusts" must {
```
Comment (adamconder) [inline test/controllers/actions/AuthActionSpec.scala]: @jassalrichy Minor change in that the Agent has enrolled for AgentServicesAccount not Trusts.

File: test/controllers/actions/AuthActionSpec.scala [hunk 4]
```diff
@@ -35,8 +35,76 @@ class AuthActionSpec extends SpecBase {
...
+        val bodyParsers = application.injector.instanceOf[BodyParsers.Default]
+        val authAction = new AuthenticatedIdentifierAction(new FakeAuthConnector(authRetrievals(agentAffinityGroup, agentEnrolment)), frontendAppConfig, bodyParsers)
+        val controller = new Harness(authAction)
+        val result = controller.onPageLoad()(fakeRequest)
+        status(result) mustBe OK
+      }
+
+    }
+
+    "Org user with no enrolled in Trusts" must {
```
Comment (adamconder) [inline test/controllers/actions/AuthActionSpec.scala]: An Organisation doesn't enrol in Trusts in order to access the service.

File: app/config/FrontendAppConfig.scala [hunk 5]
```diff
@@ -36,6 +37,7 @@ class FrontendAppConfig @Inject() (configuration: Configuration) {
   val reportAProblemNonJSUrl = s"$contactHost/contact/problem_reports_nonjs?service=$contactFormServiceIdentifier"
   val betaFeedbackUrl = s"$contactHost/contact/beta-feedback"
   val betaFeedbackUnauthenticatedUrl = s"$contactHost/contact/beta-feedback-unauthenticated"
+  val agentServiceRegistrationUrl = s"$agentServicesFrontendHost/agent-subscription/start?continue=/trusts-registration/trust-registered-online"
```
Comment (adamconder) [inline app/config/FrontendAppConfig.scala]: `/trust-registered-online` This endpoint doesn't exist in the microservice, what should this be?

File: app/config/FrontendAppConfig.scala [hunk 6]
```diff
@@ -27,6 +27,7 @@ class FrontendAppConfig @Inject() (configuration: Configuration) {
 
   private val contactHost = configuration.get[String]("contact-frontend.host")
   private val contactFormServiceIdentifier = "trustsfrontend"
+  private val agentServicesFrontendHost = configuration.get[String]("microservice.services.agent-subscription-frontend.host")
```
Comment (adamconder) [inline app/config/FrontendAppConfig.scala]: Can we please load this as a `.get[Service]` to concatenate a host and port correctly from config.

**Pull request merged**